src/object.c

```diff
@@ -1246,21 +1246,26 @@ struct redisMemOverhead *getMemoryOverheadData(void) {
 
     for (j = 0; j < server.dbnum; j++) {
         redisDb *db = server.db+j;
-        unsigned long long keyscount = dbSize(db);
+        unsigned long long keyscount = dbSize(db, MAIN_DICT);
         if (keyscount==0) continue;
 
         mh->total_keys += keyscount;
         mh->db = zrealloc(mh->db,sizeof(mh->db[0])*(mh->num_dbs+1));
         mh->db[mh->num_dbs].dbid = j;
 
         mem = keyscount * dictEntryMemUsage() +
-              dbSlots(db) * sizeof(dictEntry*) +
+              dbSlots(db, MAIN_DICT) * sizeof(dictEntry*) +
               keyscount * sizeof(robj) +
               db->dict_count * sizeof(dict);
         mh->db[mh->num_dbs].overhead_ht_main = mem;
         mem_total+=mem;
 
-        mem = dictMemUsage(db->expires);
+        unsigned long long expire_keys_count = dbSize(db, EXPIRE_DICT);
+        if (expire_keys_count == 0) continue;
+        
+        mem = expire_keys_count * dictEntryMemUsage() +
+              dbSlots(db, EXPIRE_DICT) * sizeof(dictEntry*) +
+              db->dict_count * sizeof(dict);
```

Comment:
i think i already commented on that in a previous review.
this code assumes things on the dict internals. we need to move that part of logic to dict.c